File: extras/Python/tests/loopback_scenarios.py
"""Shared host-to-embedded loopback scenarios.

These mirror ``test/CSharp/CommandMessenger.IntegrationTests/LoopbackScenariosBase.cs``
and are intentionally transport-agnostic.
"""
from __future__ import annotations

import math
import struct
import threading

from cmd_messenger import CmdMessenger, SendCommand, escaping

from .firmware_simulator import LoopbackCommand


ACK_TIMEOUT_MS = 2000
BOOT_TIMEOUT_MS = 5000


def wait_for_boot_ack(messenger: CmdMessenger, timeout_ms: int = BOOT_TIMEOUT_MS) -> None:
    ready = threading.Event()
    boot_commands = []

    messenger.attach(LoopbackCommand.ACKNOWLEDGE, lambda cmd: (boot_commands.append(cmd), ready.set()))
    assert ready.wait(timeout_ms / 1000), "Firmware did not send boot ack"
    assert boot_commands[0].cmd_id == LoopbackCommand.ACKNOWLEDGE


def assert_ping_returns_pong(messenger: CmdMessenger) -> None:
    reply = messenger.send_command(
        SendCommand(LoopbackCommand.PING, ack_cmd_id=LoopbackCommand.PONG, timeout=ACK_TIMEOUT_MS)
    )

    assert reply.ok
    assert reply.cmd_id == LoopbackCommand.PONG
    assert reply.read_string_arg() == "pong"


def assert_echo_string_round_trips(messenger: CmdMessenger) -> None:
    for text in ["hello", "Hello, World!", "  spaced  ", "special chars: !@#$%^&*()"]:
        _assert_echo_string(messenger, text)


def assert_echo_string_with_special_chars_round_trips(messenger: CmdMessenger) -> None:
    for text in ["contains, comma", "contains; semicolon", "contains/ slash", "a, b; c/ d"]:
        _assert_echo_string(messenger, text)


def assert_add_floats_returns_sum_and_difference(messenger: CmdMessenger) -> None:
    for first, second in [(0.0, 0.0), (1.0, 2.0), (-5.0, 10.0), (3.14, 2.71), (1000.5, -500.25)]:
        command = SendCommand(
            LoopbackCommand.ADD_FLOATS,
            ack_cmd_id=LoopbackCommand.ADD_FLOATS_RESULT,
            timeout=ACK_TIMEOUT_MS,
        )
        command.add_argument(first)
        command.add_argument(second)
        reply = messenger.send_command(command)

        assert reply.ok
        assert math.isclose(reply.read_float_arg(), first + second, rel_tol=1e-5, abs_tol=1e-5)
        assert math.isclose(reply.read_float_arg(), first - second, rel_tol=1e-5, abs_tol=1e-5)


def assert_echo_int32_round_trips(messenger: CmdMessenger) -> None:
    for value in [0, 42, -1, 2**31 - 1, -(2**31)]:
        reply = messenger.send_command(
            SendCommand(
                LoopbackCommand.ECHO_INT,
                value,
                ack_cmd_id=LoopbackCommand.ECHO_INT_RESULT,
                timeout=ACK_TIMEOUT_MS,
            )
        )

        assert reply.ok
        assert reply.read_int32_arg() == value


def assert_echo_int16_round_trips(messenger: CmdMessenger) -> None:
    for value in [0, 1234, -1234, 32767, -32768]:
        reply = messenger.send_command(
            SendCommand(
                LoopbackCommand.ECHO_INT16,
                value,
                ack_cmd_id=LoopbackCommand.ECHO_INT16_RESULT,
                timeout=ACK_TIMEOUT_MS,
            )
        )

        assert reply.ok
        assert reply.read_int16_arg() == value


def assert_echo_bool_round_trips(messenger: CmdMessenger) -> None:
    for value in [True, False]:
        command = SendCommand(
            LoopbackCommand.ECHO_BOOL,
            ack_cmd_id=LoopbackCommand.ECHO_BOOL_RESULT,
            timeout=ACK_TIMEOUT_MS,
        )
        command.add_argument(value)
        reply = messenger.send_command(command)

        assert reply.ok
        assert reply.read_int32_arg() == (1 if value else 0)


def assert_echo_double_round_trips_as_float_text(messenger: CmdMessenger) -> None:
    for value in [0.0, 3.141592653589793, -1.5e-10, 1.5e10]:
        command = SendCommand(
            LoopbackCommand.ECHO_DOUBLE,
            ack_cmd_id=LoopbackCommand.ECHO_DOUBLE_RESULT,
            timeout=ACK_TIMEOUT_MS,
        )
        command.add_argument(value)
        reply = messenger.send_command(command)

        assert reply.ok
        expected = _float32_text_value(value)
        actual = reply.read_float_arg()
        tolerance = max(1e-6, abs(expected) * 1e-6)
        assert expected - tolerance <= actual <= expected + tolerance


def assert_multi_args_all_types_round_trip(messenger: CmdMessenger) -> None:
    command = SendCommand(
        LoopbackCommand.MULTI_ARGS,
        ack_cmd_id=LoopbackCommand.MULTI_ARGS_RESULT,
        timeout=ACK_TIMEOUT_MS,
    )
    command.add_argument(1234)
    command.add_argument(3.14)
    command.add_argument(escaping.escape("mixed, args; here"))
    command.add_argument(True)

    reply = messenger.send_command(command)

    assert reply.ok
    assert reply.read_int16_arg() == 1234
    assert math.isclose(reply.read_float_arg(), 3.14, rel_tol=1e-5, abs_tol=1e-5)
    assert escaping.unescape(reply.read_string_arg()) == "mixed, args; here"
    assert reply.read_int32_arg() == 1


def assert_unknown_command_triggers_error(messenger: CmdMessenger) -> None:
    received = []
    event = threading.Event()
    messenger.attach(LoopbackCommand.ERROR, lambda cmd: (received.append(cmd), event.set()))

    messenger.send_command(SendCommand(99))

    assert event.wait(ACK_TIMEOUT_MS / 1000)
    assert received[0].cmd_id == LoopbackCommand.ERROR


def assert_repeated_commands_all_round_trip(messenger: CmdMessenger) -> None:
    for index in range(20):
        reply = messenger.send_command(
            SendCommand(
                LoopbackCommand.ECHO_INT,
                index,
                ack_cmd_id=LoopbackCommand.ECHO_INT_RESULT,
                timeout=ACK_TIMEOUT_MS,
            )
        )

        assert reply.ok
        assert reply.read_int32_arg() == index


def _assert_echo_string(messenger: CmdMessenger, text: str) -> None:
    reply = messenger.send_command(
        SendCommand(
            LoopbackCommand.ECHO,
            escaping.escape(text),
            ack_cmd_id=LoopbackCommand.ECHO_RESULT,
            timeout=ACK_TIMEOUT_MS,
        )
    )

    assert reply.ok
    assert escaping.unescape(reply.read_string_arg()) == text


def _float32_text_value(value: float) -> float:
    return struct.unpack("<f", struct.pack("<f", value))[0]

File: extras/TypeScript/tests/loopbackScenarios.ts
import { expect } from 'vitest';
import { CmdMessenger } from '../src/cmdMessenger.js';
import { escape, unescape } from '../src/escaping.js';
import { ReceivedCommand } from '../src/receivedCommand.js';
import { SendCommand } from '../src/sendCommand.js';
import { LoopbackCommand } from './loopbackCommands.js';

export const ACK_TIMEOUT_MS = 2000;
export const BOOT_TIMEOUT_MS = 5000;

export interface LoopbackScenarioOptions {
  ackTimeoutMs?: number;
}

export type LoopbackScenario = {
  name: string;
  run: (messenger: CmdMessenger, options?: LoopbackScenarioOptions) => Promise<void>;
};

const floatCases: Array<readonly [number, number]> = [
  [0, 0],
  [1, 2],
  [-5, 10],
  [3.14, 2.71],
  [1000.5, -500.25]
];

function timeout(options?: LoopbackScenarioOptions): number {
  return options?.ackTimeoutMs ?? ACK_TIMEOUT_MS;
}

export function waitForCommand(
  messenger: CmdMessenger,
  cmdId: number,
  timeoutMs: number
): Promise<ReceivedCommand> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(`Timed out waiting for command ${cmdId} after ${timeoutMs} ms`));
    }, timeoutMs);

    messenger.attach(cmdId, (command) => {
      clearTimeout(timer);
      resolve(command);
    });
  });
}

export async function waitForBootAck(
  messenger: CmdMessenger,
  timeoutMs = BOOT_TIMEOUT_MS
): Promise<ReceivedCommand> {
  const boot = await waitForCommand(messenger, LoopbackCommand.Acknowledge, timeoutMs);
  expect(boot.cmdId).toBe(LoopbackCommand.Acknowledge);
  return boot;
}

export async function assertPingReturnsPong(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  const reply = await messenger.sendCommand(
    new SendCommand(LoopbackCommand.Ping).withAck(LoopbackCommand.Pong, timeout(options))
  );

  expect(reply.ok).toBe(true);
  expect(reply.cmdId).toBe(LoopbackCommand.Pong);
  expect(reply.readStringArg()).toBe('pong');
}

export async function assertEchoStringRoundTrips(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const text of ['hello', 'Hello, World!', '  spaced  ', 'special chars: !@#$%^&*()']) {
    await assertEchoString(messenger, text, options);
  }
}

export async function assertEchoStringWithSpecialCharsRoundTrips(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const text of ['contains, comma', 'contains; semicolon', 'contains/ slash', 'a, b; c/ d']) {
    await assertEchoString(messenger, text, options);
  }
}

export async function assertAddFloatsReturnsSumAndDifference(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const [first, second] of floatCases) {
    const command = new SendCommand(LoopbackCommand.AddFloats)
      .addArgument(first)
      .addArgument(second)
      .withAck(LoopbackCommand.AddFloatsResult, timeout(options));
    const reply = await messenger.sendCommand(command);

    expect(reply.ok).toBe(true);
    expect(reply.readFloatArg()).toBeCloseTo(first + second, 4);
    expect(reply.readFloatArg()).toBeCloseTo(first - second, 4);
  }
}

export async function assertEchoInt32RoundTrips(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const value of [0, 42, -1, 2 ** 31 - 1, -(2 ** 31)]) {
    const reply = await messenger.sendCommand(
      new SendCommand(LoopbackCommand.EchoInt, value).withAck(LoopbackCommand.EchoIntResult, timeout(options))
    );

    expect(reply.ok).toBe(true);
    expect(reply.readInt32Arg()).toBe(value);
  }
}

export async function assertEchoInt16RoundTrips(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const value of [0, 1234, -1234, 32767, -32768]) {
    const reply = await messenger.sendCommand(
      new SendCommand(LoopbackCommand.EchoInt16, value).withAck(LoopbackCommand.EchoInt16Result, timeout(options))
    );

    expect(reply.ok).toBe(true);
    expect(reply.readInt16Arg()).toBe(value);
  }
}

export async function assertEchoBoolRoundTrips(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const value of [true, false]) {
    const command = new SendCommand(LoopbackCommand.EchoBool)
      .addArgument(value)
      .withAck(LoopbackCommand.EchoBoolResult, timeout(options));
    const reply = await messenger.sendCommand(command);

    expect(reply.ok).toBe(true);
    expect(reply.readInt32Arg()).toBe(value ? 1 : 0);
  }
}

export async function assertEchoDoubleRoundTripsAsFloatText(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const value of [0, Math.PI, -1.5e-10, 1.5e10]) {
    const command = new SendCommand(LoopbackCommand.EchoDouble)
      .addArgument(value)
      .withAck(LoopbackCommand.EchoDoubleResult, timeout(options));
    const reply = await messenger.sendCommand(command);

    expect(reply.ok).toBe(true);
    const expected = float32Value(value);
    const actual = reply.readFloatArg();
    const tolerance = Math.max(1e-6, Math.abs(expected) * 1e-6);
    expect(actual).toBeGreaterThanOrEqual(expected - tolerance);
    expect(actual).toBeLessThanOrEqual(expected + tolerance);
  }
}

export async function assertMultiArgsAllTypesRoundTrip(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  const command = new SendCommand(LoopbackCommand.MultiArgs)
    .addArgument(1234)
    .addArgument(3.14)
    .addArgument(escape('mixed, args; here'))
    .addArgument(true)
    .withAck(LoopbackCommand.MultiArgsResult, timeout(options));

  const reply = await messenger.sendCommand(command);

  expect(reply.ok).toBe(true);
  expect(reply.readInt16Arg()).toBe(1234);
  expect(reply.readFloatArg()).toBeCloseTo(3.14, 4);
  expect(unescape(reply.readStringArg())).toBe('mixed, args; here');
  expect(reply.readInt32Arg()).toBe(1);
}

export async function assertUnknownCommandTriggersError(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  const error = waitForCommand(messenger, LoopbackCommand.Error, timeout(options));
  await messenger.sendCommand(new SendCommand(99));
  const reply = await error;

  expect(reply.cmdId).toBe(LoopbackCommand.Error);
}

export async function assertRepeatedCommandsAllRoundTrip(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (let index = 0; index < 20; index += 1) {
    const reply = await messenger.sendCommand(
      new SendCommand(LoopbackCommand.EchoInt, index).withAck(LoopbackCommand.EchoIntResult, timeout(options))
    );

    expect(reply.ok).toBe(true);
    expect(reply.readInt32Arg()).toBe(index);
  }
}

export const loopbackScenarioCases: LoopbackScenario[] = [
  { name: 'ping returns pong', run: assertPingReturnsPong },
  { name: 'echo string round trips', run: assertEchoStringRoundTrips },
  { name: 'escaped string round trips', run: assertEchoStringWithSpecialCharsRoundTrips },
  { name: 'add floats returns sum and difference', run: assertAddFloatsReturnsSumAndDifference },
  { name: 'echo int32 round trips', run: assertEchoInt32RoundTrips },
  { name: 'echo int16 round trips', run: assertEchoInt16RoundTrips },
  { name: 'echo bool round trips', run: assertEchoBoolRoundTrips },
  { name: 'echo double round trips as float text', run: assertEchoDoubleRoundTripsAsFloatText },
  { name: 'multi args all types round trip', run: assertMultiArgsAllTypesRoundTrip },
  { name: 'unknown command triggers error', run: assertUnknownCommandTriggersError },
  { name: 'repeated commands all round trip', run: assertRepeatedCommandsAllRoundTrip }
];

export async function assertAllLoopbackScenarios(
  messenger: CmdMessenger,
  options?: LoopbackScenarioOptions
): Promise<void> {
  for (const scenario of loopbackScenarioCases) {
    await scenario.run(messenger, options);
  }
}

async function assertEchoString(
  messenger: CmdMessenger,
  text: string,
  options?: LoopbackScenarioOptions
): Promise<void> {
  const reply = await messenger.sendCommand(
    new SendCommand(LoopbackCommand.Echo, escape(text)).withAck(LoopbackCommand.EchoResult, timeout(options))
  );

  expect(reply.ok).toBe(true);
  expect(unescape(reply.readStringArg())).toBe(text);
}

function float32Value(value: number): number {
  const buffer = new ArrayBuffer(4);
  const view = new DataView(buffer);
  view.setFloat32(0, value, true);
  return view.getFloat32(0, true);
}
